File: app/src/main/java/com/i14yokoro/mldpterminal/EscapeSequence.java
package com.i14yokoro.mldpterminal;

/**
 * ANSIのエスケープシーケンスと同じ動作をする
 */
public class EscapeSequence {
    private TermDisplay termDisplay;

    /**
     * @param termDisplay 表示画面の情報
     */
    EscapeSequence(TermDisplay termDisplay){
        this.termDisplay = termDisplay;
    }

    /**
     * @param n 移動する量
     */
    public void moveRight(int n){
        if (n >= termDisplay.getDisplayRowSize()) {
            n = termDisplay.getDisplayRowSize() - 1;
        }

        if(termDisplay.getCursorX() + n < termDisplay.getRowLength(getSelectRowIndex())) {
            moveCursorX(n);
        } else {
            int move = n;
            int add;
            if(termDisplay.getCursorX() + n >= termDisplay.getDisplayRowSize()){
                add = termDisplay.getDisplayRowSize() - termDisplay.getRowLength(getSelectRowIndex());
                move = termDisplay.getDisplayRowSize() - termDisplay.getCursorX();
            } else {
                add = termDisplay.getCursorX() + n - termDisplay.getRowLength(getSelectRowIndex());
            }
            addBlank(add);
            moveCursorX(move);
        }
    }

    /**
     * @param n 移動する量
     */
    public void moveLeft(int n){
        if (n >= termDisplay.getDisplayRowSize()) {
            n = termDisplay.getDisplayRowSize() - 1;
        }

        if(termDisplay.getCursorX() - n >= 0){
            moveCursorX(-n);
        }
    }

    /**
     * @param n 移動する量
     */
    public void moveUp(int n){
        if (n >= termDisplay.getDisplayColumnSize()) {
            n = termDisplay.getDisplayColumnSize() - 1;
        }

        if(termDisplay.getCursorY() - n < 0){ //画面外にでる
            termDisplay.setCursorY(0);
        } else {
            moveCursorY(-n);
        }
        int rowLength = termDisplay.getRowLength(getSelectRowIndex());

        if(rowLength < termDisplay.getCursorX()){
            int add = termDisplay.getCursorX() - rowLength + 1;
            addBlank(add);
        }
    }

    /**
     * @param n 移動する量
     */
    public void moveDown(int n){
        if (n >= termDisplay.getDisplayColumnSize()) {
            n = termDisplay.getDisplayColumnSize() - 1;
        }

        if(termDisplay.getCursorY() + n >= termDisplay.getDisplaySize()) { //移動先が一番下の行を超える場合
            if(termDisplay.getCursorY() + n >= termDisplay.getDisplayColumnSize()){ //ディスプレイサイズを超える場合
                termDisplay.setCursorY(termDisplay.getDisplayColumnSize() - 1);
            } else {
                moveCursorY(n);
            }
            int move = termDisplay.getCursorY() - termDisplay.getDisplaySize();
            for (int i = 0; i <= move; i++){
                termDisplay.addEmptyRow();
            }
        }else { //移動先が一番下の行を超えない
            if (termDisplay.getCursorY() + n > termDisplay.getDisplaySize()) { //ディスプレイサイズを超える場合
                termDisplay.setCursorY(termDisplay.getDisplayColumnSize() - 1);
            } else {
                moveCursorY(n);
            }
        }
        //移動先の文字数がcursorXより小さい
        if(termDisplay.getRowLength(getSelectRowIndex()) < termDisplay.getCursorX()){
            int add = termDisplay.getCursorX() - termDisplay.getRowLength(getSelectRowIndex());
            addBlank(add);
        }
    }

    /**
     * @param n 追加する空白の量
     */
    private void addBlank(int n){
        int x = termDisplay.getRowLength(getSelectRowIndex())-1; //現在の行の最後のindexを取得
        if(x < 0){
            x = 0;
        }
        for (int i = 0; i < n; i++){
            termDisplay.insertTextItem(x, getSelectRowIndex(),' ', termDisplay.getDefaultColor());
            x = termDisplay.getRowLength(getSelectRowIndex())-1;
        }
    }

    /**
     * @param n 移動する量
     */
    public void moveUpToRowLead(int n){
        termDisplay.setCursorX(0);
        moveUp(n);
    }

    /**
     * @param n 移動する量
     */
    public void moveDownToRowLead(int n){
        termDisplay.setCursorX(0);
        moveDown(n);
    }

    /**
     * @param n 移動する量
     */
    public void moveSelection(int n){
        termDisplay.setCursorX(0);
        moveRight(n-1);
    }

    /**
     * @param n 縦に移動する量(0 < n)
     * @param m 横に移動する量(0 < m)
     */
    public void moveSelection(int n, int m){ //n,mは1~
        if(n > termDisplay.getDisplayColumnSize()){
            n = termDisplay.getDisplayColumnSize();
        }
        if (m > termDisplay.getDisplayRowSize()){
            m = termDisplay.getDisplayRowSize();
        }
        termDisplay.setCursorY(0);
        termDisplay.setCursorX(0);
        moveDown(n-1);
        moveRight(m-1);
    }

    /**
     * 画面消去
     */
    public void clearDisplay(){
        int x = termDisplay.getCursorX();
        int i = x;
        int length;
        int displaySize = termDisplay.getDisplaySize();
        for(int y = termDisplay.getCursorY(); y < displaySize; y++){
            length = termDisplay.getRowLength(y + termDisplay.getTopRow());
            for (; i < length; i++){
                if(termDisplay.getRowText(y + termDisplay.getTopRow()).length() > x) {
                    termDisplay.deleteTextItem(x, y + termDisplay.getTopRow());
                }
            }
            x = 0;
            i = 0;
        }
    }

    /**
     * @param n 画面消去の方法
     */
    public void clearDisplay(int n){
        //一番下までスクロールして消去ぽい
        if(n == 0){ //カーソルより後ろにある画面上の文字を消す
            clearDisplay();
        }

        if(n == 1){ //カーソルより前にある画面上の文字を消す
            for (int y = termDisplay.getTopRow(); y <= getSelectRowIndex(); y++){
                for (int x = 0; x < termDisplay.getRowLength(y); x++){
                    termDisplay.changeText(x, y, '\u0000');
                    if (y == getSelectRowIndex() && x == termDisplay.getCursorX()) {
                        break;
                    }
                }
            }
        }

        if(n == 2){ //全消去
            for (int y = termDisplay.getTopRow(); y < termDisplay.getDisplaySize() + termDisplay.getTopRow(); y++){
                for (int x = 0; x < termDisplay.getRowLength(y); x++){
                    termDisplay.changeTextItem(x, y, '\u0000', termDisplay.getDefaultColor());
                }
            }
        }

    }

    /**
     * @param n 行削除の方法
     */
    public void clearRow(int n){
        if(n == 0){ //カーソル以降にある文字を消す
            int del = termDisplay.getRowLength(getSelectRowIndex()) - termDisplay.getCursorX();
            for (int x = 0; x < del; x++){
                termDisplay.deleteTextItem(termDisplay.getCursorX(), getSelectRowIndex());
            }
        }

        if(n == 1){ //カーソル以前にある文字を消す
            for (int x = 0; x <= termDisplay.getCursorX(); x++){
                termDisplay.changeText(x, getSelectRowIndex(), ' ');
            }
        }

        if(n == 2){ //全消去
            int del = termDisplay.getRowLength(getSelectRowIndex());
            for (int x = 0; x < del; x++){
                termDisplay.changeText(x, getSelectRowIndex(), ' ');
            }

        }
    }

    /**
     * @param n 移動する量
     */
    public void scrollNext(int n){
        if (termDisplay.getTopRow() + n > termDisplay.getTotalColumns()) return; //一番したに空白追加？？
        termDisplay.setTopRow(termDisplay.getTopRow()+n);
    }

    /**
     * @param n 移動する量
     */
    public void scrollBack(int n){
        //一番上に空白追加で一番した削除？？？(あくまで画面上でスクロールしていると見せかけている?)
        if(termDisplay.getTopRow() - n < 0) return;
        termDisplay.setTopRow(termDisplay.getTopRow()-n);
    }

    /**
     * @param n 変化させる色
     */
    public void selectGraphicRendition(int n){ //色
        termDisplay.setColorChange(true);
        switch (n){
            case 0:
                termDisplay.setDefaultColor(0xFF000000);
                break;
            case 30:
                termDisplay.setDefaultColor(0xFF000000);
                break;
            case 31:
                termDisplay.setDefaultColor(0xFFff0000);
                break;
            case 32:
                termDisplay.setDefaultColor(0xFF008000);
                break;
            case 33:
                termDisplay.setDefaultColor(0xFFFFFF00);
                break;
            case 34:
                termDisplay.setDefaultColor(0xFF0000FF);
                break;
            case 35:
                termDisplay.setDefaultColor(0xFFFF00FF);
                break;
            case 36:
                termDisplay.setDefaultColor(0xFF00FFFF);
                break;
            case 37:
                termDisplay.setDefaultColor(0xFFFFFFFF);
                break;
            case 39:
                termDisplay.setDefaultColor(0xFF000000);
                break;
            default:
                //termDisplay.setDefaultColor(0x000000);
        }
    }

    private int getSelectRowIndex() {
        return termDisplay.getCursorY() + termDisplay.getTopRow();
    }

    /**
     * @param x カーソルをx移動させる
     */
    private void moveCursorX(int x){
        termDisplay.setCursorX(termDisplay.getCursorX() + x);
    }

    /**
     * @param y カーソルをy移動させる
     */
    private void moveCursorY(int y){
        termDisplay.setCursorY(termDisplay.getCursorY() + y);
    }

}

File: app/src/main/java/com/i14yokoro/mldpterminal/EscapeSequence.kt
package com.i14yokoro.mldpterminal

import com.i14yokoro.mldpterminal.terminalview.TerminalView.Cursor

/**
 * ANSIのエスケープシーケンスと同じ動作をする
 */
class EscapeSequence
/**
 * @param termBuffer 表示画面の情報
 */
internal constructor(private val termBuffer: TerminalBuffer) {

    private val space = ' '

    /**
     * @param n 移動量
     */
    fun moveRight(cursor: Cursor, n: Int) {
        moveCursorX(cursor, n)
    }

    /**
     * @param n 移動量
     */
    fun moveLeft(cursor: Cursor, n: Int) {
        moveCursorX(cursor, -n)
    }

    /**
     * @param n 移動量
     */
    fun moveUp(cursor: Cursor, n: Int) {
        moveCursorY(cursor, -n)
    }

    /**
     * @param n 移動量
     */
    fun moveDown(cursor: Cursor, n: Int) {
        moveCursorY(cursor, n)

        if(termBuffer.topRow + cursor.y >= termBuffer.totalLines) {
            for(i in 0 .. termBuffer.topRow + cursor.y - termBuffer.totalLines) {
                termBuffer.addRow()
            }
        }
    }

    /**
     * @param n 移動量
     */
    fun moveUpToRowLead(cursor: Cursor, n: Int) {
        cursor.x = 0
        moveUp(cursor, n)
    }

    /**
     * @param n 移動量
     */
    fun moveDownToRowLead(cursor: Cursor, n: Int) {
        cursor.x = 0
        moveDown(cursor, n)
    }

    /**
     * @param n 移動量
     */
    fun moveCursor(cursor: Cursor, n: Int) {
        cursor.x = 0
        moveRight(cursor, n-1)
    }

    /**
     * @param n 縦の移動量(0 < n)
     * @param m 横の移動量(0 < m)
     */
    fun moveCursor(cursor: Cursor, n: Int, m: Int) { //n,mは1~
        if(n-1 > cursor.y) {
            cursor.y = 0
            moveDown(cursor, n-1)
        } else {
            cursor.y = n-1
        }

        cursor.x = m-1
    }

    /**
     * 画面消去
     */
    private fun clearDisplay(cursor: Cursor) {
        for (x in cursor.x until termBuffer.screenColumnSize){
            termBuffer.setText(x, termBuffer.topRow + cursor.y, space)
        }
        for (y in cursor.y+1 until termBuffer.displayedLines) {
            for (x in 0 until termBuffer.screenColumnSize){
                termBuffer.setText(x, y, space)
            }
        }
    }

    /**
     * @param n 画面消去の方法
     */
    fun clearDisplay(cursor: Cursor, n: Int) {
        //一番下までスクロールして消去ぽい
        if (n == 0) { //カーソルより後ろにある画面上の文字を消す
            clearDisplay(cursor)
        }

        if (n == 1) { //カーソルより前にある画面上の文字を消す
            for (y in termBuffer.topRow .. termBuffer.topRow + cursor.y) {
                for (x in 0 until termBuffer.screenColumnSize) {
                    if (y == termBuffer.topRow + cursor.y && x == cursor.x) {
                        break
                    }
                    termBuffer.setText(x, y, space)
                }
            }
        }

        if (n == 2) { //全消去
            for (y in termBuffer.topRow until termBuffer.displayedLines + termBuffer.topRow) {
                for (x in 0 until termBuffer.screenColumnSize) {
                    termBuffer.setText(x, y, space)
                }
            }
        }
    }

    /**
     * @param n 行削除の方法
     */
    fun clearRow(cursor: Cursor, n: Int) {
        if (n == 0) { //カーソル以降にある文字を消す
            for (x in cursor.x until termBuffer.screenColumnSize) {
                termBuffer.setText(x, termBuffer.topRow + cursor.y, space)
            }
        }

        if (n == 1) { //カーソル以前にある文字を消す
            for (x in 0 until cursor.x) {
                termBuffer.setText(x, termBuffer.topRow + cursor.y, space)
            }
        }

        if (n == 2) { //全消去
            for (x in 0 until termBuffer.screenColumnSize) {
                termBuffer.setText(x, termBuffer.topRow + cursor.y, space)
            }
        }
    }

    /**
     * @param n 移動する量
     */
    fun scrollNext(n: Int) {
        if (termBuffer.topRow + n > termBuffer.totalLines) return  //一番したに空白追加？？
        termBuffer.topRow = termBuffer.topRow + n
    }

    /**
     * @param n 移動する量
     */
    fun scrollBack(n: Int) {
        //一番上に追加で一番した削除？？？(あくまで画面上でスクロールしていると見せかけている?)
        if (termBuffer.topRow - n < 0) return
        termBuffer.topRow = termBuffer.topRow - n
    }

    /**
     * @param n 変化させる色
     */
    fun selectGraphicRendition(n: Int) { //色
        termBuffer.isColorChange = true
        when (n) {
            0, 39, 30 -> termBuffer.charColor = 0xFF000000.toInt()
            31 -> termBuffer.charColor = 0xFFff0000.toInt()
            32 -> termBuffer.charColor = 0xFF008000.toInt()
            33 -> termBuffer.charColor = 0xFFFFFF00.toInt()
            34 -> termBuffer.charColor = 0xFF0000FF.toInt()
            35 -> termBuffer.charColor = 0xFFFF00FF.toInt()
            36 -> termBuffer.charColor = 0xFF00FFFF.toInt()
            37 -> termBuffer.charColor = 0xFFFFFFFF.toInt()
        }
    }

    /**
     * @param x カーソルをx移動させる
     */
    private fun moveCursorX(cursor: Cursor, x: Int) {
        cursor.x += x
    }

    /**
     * @param y カーソルをy移動させる
     */
    private fun moveCursorY(cursor: Cursor, y: Int) {
        cursor.y +=  y
    }

}
